src/utils/size-observer.js

```diff
@@ -0,0 +1,142 @@
+
+/** @typedef {function(../layout-rect.LayoutSizeDef)} */
+let ContentSizeObserverCallbackDef;
+
+/** @const {!WeakMap<!Window, !ResizeObserver>} */
+const observers = new WeakMap();
+
+/** @const {!WeakMap<!Element, !ContentSizeObserverCallbackDef>} */
+const targetObserverListMap = new WeakMap();
+
+/** @const {!WeakMap<!Element, !../layout-rect.LayoutSizeDef>} */
+const targetSizeListMap = new WeakMap();
+
+/**
+ * @param {!Element} element
+ * @param {!ContentSizeObserverCallbackDef} callback
+ */
+export function observeContentSize(element, callback) {
+  const win = element.ownerDocument.defaultView;
+  if (!win) {
+    return;
+  }
+  let callbacks = targetObserverListMap.get(element);
+  if (!callbacks) {
+    callbacks = [];
+    targetObserverListMap.set(element, callbacks);
+    getObserver(win).observe(element);
+  }
+  if (pushIfNotExist(callbacks, callback)) {
+    const size = targetSizeListMap.get(element);
+    if (size) {
+      setTimeout(() => callCallbackNoInline(callback, size));
+    }
+  }
+}
+
+/**
+ * @param {!Element} element
+ * @param {!ObserverCallbackDef} callback
+ */
+export function unobserveContentSize(element, callback) {
+  const callbacks = targetObserverListMap.get(element);
+  if (!callbacks) {
+    return;
+  }
+  removeItem(callbacks, callback);
+  if (callbacks.length == 0) {
+    targetObserverListMap.delete(element);
+    targetSizeListMap.delete(element);
+    const win = element.ownerDocument.defaultView;
+    if (win) {
+      getObserver(win).unobserve(element);
+    }
+  }
+}
+
+/**
+ * @param {!Element} element
+ * @return {!Promise<!../layout-rect.LayoutSizeDef>}
+ */
+export function measureContentSize(element) {
+  return new Promise((resolve) => {
+    const onSize = (size) => {
+      resolve(size);
+      unobserveContentSize(element, onSize);
+    };
+    observeContentSize(element, onSize);
+  });
+}
+
+/**
+ * @param {!Window} win
+ * @return {!ResizeObserver}
+ */
+function getObserver(win) {
+  let observer = observers.get(win);
+  if (!observer) {
+    observer = new win.ResizeObserver(processEntries);
+    observers.set(win, observer);
+  }
+  return observer;
+}
+
+/**
+ * @param {!Array<!ResizeObserverEntry>} entries
+ */
+function processEntries(entries) {
+  for (let i = entries.length - 1; i >= 0; i--) {
+    const {target, contentRect} = entries[i];
+    let seenTarget = false;
+    for (let j = i + 1; j < entries.length; j++) {
+      if (entries[j].target == target) {
+        seenTarget = true;
+        break;
+      }
+    }
+    if (seenTarget) {
+      continue;
+    }
+    const callbacks = targetObserverListMap.get(target);
+    if (!callbacks) {
+      continue;
+    }
+    const {width, height} = contentRect;
+    /** @type {../layout-rect.LayoutSizeDef} */
+    const size = {width, height};
+    targetSizeListMap.set(target, size);
+    for (let k = 0; k < callbacks.length; k++) {
+      callCallbackNoInline(callbacks[k], size);
```

Comment:
Is this so that a single error in a callback doesn't cause the rest to skip execution?
If yes, I should probably add this to the `measureIntersection` helper

Reply:
Correct. But I don't think you have that problem in `measureIntersection`. AFAIR you only resolve promises there, which execute their own `try/catch` for `then` callbacks.